stock_reserve_sale/models/sale.py

```diff
@@ -0,0 +1,251 @@
+                    )
+                    % (order.name)
+                )
+        return super().unlink()
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    def _get_line_rule(self):
+        """Get applicable rule for this product
+
+        Reproduce get suitable rule from procurement
+        to predict source location"""
+        StockRule = self.env["stock.rule"]
+        product = self.product_id
+        product_route_ids = [
+            x.id for x in product.route_ids + product.categ_id.total_route_ids
+        ]
+        rules = StockRule.search(
+            [("route_id", "in", product_route_ids)],
+            order="route_sequence, sequence",
+            limit=1,
+        )
+
+        if not rules:
+            warehouse = self.order_id.warehouse_id
+            wh_routes = warehouse.route_ids
+            wh_route_ids = [route.id for route in wh_routes]
+            domain = [
+                "|",
+                ("warehouse_id", "=", warehouse.id),
+                ("warehouse_id", "=", False),
+                ("route_id", "in", wh_route_ids),
+            ]
+
+            rules = StockRule.search(domain, order="route_sequence, sequence")
+
+        if rules:
+            fields.first(rules)
+        return False
+
+    def _get_procure_method(self):
+        """Get procure_method depending on product routes"""
+        rule = self._get_line_rule()
+        if rule:
+            return rule.procure_method
+        return False
+
+    @api.depends("state", "product_id.route_ids", "product_id.type")
+    def _compute_is_stock_reservable(self):
+        for line in self:
+            reservable = False
+            if (
+                not (
+                    line.state not in ("draft", "sent")
+                    or line._get_procure_method() == "make_to_order"
+                    or not line.product_id
+                    or line.product_id.type == "service"
+                )
+                and not line.reservation_ids
+            ):
+                reservable = True
+            line.is_stock_reservable = reservable
+
+    @api.depends("order_id.state", "reservation_ids")
+    def _compute_is_readonly(self):
+        for line in self:
+            line.is_readonly = (
+                len(line.reservation_ids) > 0 or line.order_id.state != "draft"
+            )
+
+    reservation_ids = fields.One2many(
+        "stock.reservation", "sale_line_id", string="Stock Reservation", copy=False
+    )
+    is_stock_reservable = fields.Boolean(
+        compute="_compute_is_stock_reservable", readonly=True, string="Can be reserved"
+    )
+    is_readonly = fields.Boolean(compute="_compute_is_readonly", store=False)
+
+    def release_stock_reservation(self):
+        reserv_ids = [reserv.id for line in self for reserv in line.reservation_ids]
+        reservations = self.env["stock.reservation"].browse(reserv_ids)
+        reservations.release_reserve()
```

Comment:
```suggestion
        self.reservation_ids.release_reserve()
```

Reply:
> 

@hparfr Are you sure? We run similar and we were sure we needed a rebrowse here.

Reply:
Not; I didn't run the code.

Did you investigate why you need to rebrowse here ? Does it need some cache invalidation somewhere else ?

Reply:
This code doesn't look like v14 code. 

As you know, back in the days, the ORM was not as performant as now and chained expressions were not possible. We had to write *for in a for* (like the hardtoread double `for` in the comprehension list above) or `mapped()` chain ( *somerecord.mapped('somefield_id').mapped('someother_ids')...*) it's not needed anymore.

In theory, we can expect the same result with both version of the code.

Reply:
I disagree. Both with the example and the conclusion. This isn't special to v14, or even as far back as v6 you could use that syntax albeit with a few extra args in the function. I think where you have a migration, in review it is especially important to apply the principles of Chesterton's fence. This module was written by an accomplished dev, it went through multiple years of reviews, it deliberately included a rebrowse, and until we know why that rebrowse was there, and whether it is still required, it would seem extremely risky to say "In theory they would do the same" because we know they don't always, hence why sometimes we rebrowse.

Reply:
There is no advance in this discussion. 

I *suggest* to rewrite this part in more recent style to improve the code quality: readability, aligned with other part of the v14 code, and performance.

You says in v13 the rebrowse was necessary. I trust you for the v13 code. I did not run it.

But since we both did not run this code in v14, we can't decide if the rebrowse is still really necessary anymore.

Reply:
I'm not saying you are wrong, but I know for me personally, if I see a rebrowse, that is an unusual thing to do in any version. I would want to know why it was there in the first place before suggesting it be removed. I even tried to find out, and that isn't easy with the way we port commits and that it isn't noted anywhere.  However, I just saw, we are using #1347 in a v14 production instance, which does have the rebrowse, for some months now. And still I don't know if it is necessary, but I know that it works and is subtly different to the suggestion.  We have some really bad instances lately of OCA modules breaking on seemingly minor changes and refactoring.